cosigner/authcosigner.go

```diff
@@ -0,0 +1,224 @@
+package cosigner
+
+import (
+	"crypto"
+	"crypto/hmac"
+	"crypto/rand"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/cosigner/msgs"
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type AuthCosigner struct {
+	Cosigner
+	Issuer       string
+	KeyID        string
+	authIdIter   atomic.Uint64
+	hmacKey      []byte
+	AuthStateMap map[string]*AuthState
+	AuthCodeMap  map[string]string
```

Comment:
For `AuthCodeMap` we only write to it once and the key we use is randomly chosen so

Reply:
> This assumption stems from developers viewing the table[key] syntax used for map accesses and misinterpreting them to be accessing disjoint elements. However, a map (hash table), unlike an array or a slice, is a sparse data structure, and accessing one element might result in accessing another element; if during the same process another insertion/deletion happens, it will modify the sparse data structure and cause a data race

maps in go are gross

Reply:
Looking into this more, I am just going to use RWMutex around AuthStateMap and AuthCodeMap. Anyone using this an actual MFACosigner should replace AuthStateMap and AuthCodeMap with a database.

Reply:
Fixed, but doublecheck my locking code.

I ended up completely rewriting this by adding a AuthStateStore interface and then creating a inmemory AuthStateStore which has locks on both maps.

I want to revisit this after we have a database backed production ready Cosigner to ensure the AuthStore interface I designed here actually exposes the right functions. Should it also expose the user or session maps in the example MFACosigner? I don't know and it is hard to know without a good database backed implementation.

Reply:
Locking code lgtm. I think if you run with the `-race` flag when running unit tests, as long as the unit tests touch code that involve the maps, we will find out if there are any race conditions still

Reply:
> I want to revisit this after we have a database backed production ready Cosigner to ensure the AuthStore interface I designed here actually exposes the right functions. Should it also expose the user or session maps in the example MFACosigner? I don't know and it is hard to know without a good database backed implementation.

Yeah let's wait for now. Seems good to me right now though. It might slightly change later if we think we need to expose some more things